### PR TITLE
feat(#3700): push-to-talk hold gesture and Ctrl+Shift+D keyboard shortcut for dictation

### LINKED ISSUE #3700: Add reliable dictation mode / push-to-talk option
## Feature Description

Add a reliable dictation mode for Hermes WebUI where recording continues until the user explicitly stops it, instead of ending automatically when the browser thinks speech has finished.

## Motivation

The current composer mic can stop while the user is still speaking, especially if they pause mid-sentence, hesitate, or speak in longer natural phrases. This makes dictation frustrating for longer prompts.

The practical issue is that browser speech recognition may decide an utterance has ended and stop early. Users who want reliable long-form dictation need a mode that behaves more like a normal recorder: start recording, keep recording, then stop and transcribe when the user chooses.

## Proposed Solution

Add a user-facing setting for dictation behaviour, for example:

1. **Browser dictation** — current fast browser speech-recognition mode.
2. **Reliable dictation** — record audio until the user taps Stop, then transcribe with the server-side STT path.
3. **Push-to-talk** — press/hold the mic or a keyboard shortcut to record; release to stop and transcribe.

This could be exposed in Settings near the existing voice/raw-audio options, and possibly default to reliable dictation for users who have server-side STT available.

## Alternatives Considered

- Keep relying on browser speech recognition, but set it to continuous mode. This may help in some browsers, but support is inconsistent and the browser can still end sessions unexpectedly.
- Use the existing hidden localStorage workaround to force MediaRecorder. This is useful for debugging, but it is not discoverable or user-friendly.

## Environment / Notes

Observed on Hermes WebUI using the composer mic. Server-side transcription via local STT works, but the default dictation path uses browser speech recognition first, so premature stops can happen client-side before audio ever reaches `/api/transcribe`.

### PR DESCRIPTION
**Pull request by rodboev**
## Thinking Path

- The open issue asks for broader dictation behavior, but the first mergeable slice is still the existing composer mic path: keep quick click, add hold-to-talk, and avoid shipping a browser-reserved global shortcut that fights the page.
- The current mic controller already owns SpeechRecognition, MediaRecorder, and shared stop behavior through `_stopMic()`, so pointer, keyboard, and command-triggered dictation need to converge on that same owner instead of splitting across parallel handlers.
- The review cycle exposed hidden-control and pending-permission lifecycle edges, so the final branch hardens those paths too, keeping the feature small while making the mic state machine safe to ship.

## What Changed

- `static/boot.js`: extracted shared mic start and toggle helpers, replaced the inline click owner with pointer hold and release wiring, and added start-sequencing plus per-capture cleanup so stale permission prompts or recorder callbacks cannot clobber a newer live capture.
- `static/commands.js`: aligned `/voice` with the same hidden-mic visibility contract before it triggers the mic button programmatically.
- `tests/test_issue3700_push_to_talk.py`: added focused source-contract coverage for the shared mic owner, hidden-mic guards, `/voice` alignment, pending-start lifecycle checks, and the reserved-chord regression.
- `tests/test_issue4571_mic_insecure_origin.py`: updated the insecure-origin regression to follow the extracted `_startMicCapture(...)` helper instead of the removed inline click handler marker.

## Why It Matters

Dictation gets a real push-to-talk gesture without breaking the current click-toggle workflow, and the mic controller now survives permission-prompt timing, hidden-control edge cases, and browser-level shortcut collisions that would otherwise make the feature unreliable.

## Verification

- `pytest tests/test_issue3700_push_to_talk.py tests/test_issue1488_composer_voice_buttons.py tests/test_issue4571_mic_insecure_origin.py -v --timeout=60`
- `npx eslint --no-config-lookup -c eslint.runtime-guard.config.mjs static/boot.js`
- Browser research: Google Chrome and Microsoft Edge both reserve `Ctrl/Cmd+Shift+D` for bookmarking all open tabs, so the final head keeps the hold gesture and does not bind that chord.

## Upstream

Closes #3700.

## Model Used

GPT 5.5 via Codex CLI

### CODE REVIEW

File: static/boot.js [hunk 1]
```diff
@@ -1810,6 +1918,15 @@ document.addEventListener('keydown',async e=>{
...
   }
+  if((e.metaKey||e.ctrlKey)&&e.shiftKey&&!e.altKey&&(e.key==='d'||e.key==='D')){
+    const t=e.target;
+    const isText=t&&(t.tagName==='INPUT'||t.tagName==='TEXTAREA'||t.isContentEditable);
+    if(!isText&&typeof window._toggleMicCapture==='function'){
+      e.preventDefault();
+      await window._toggleMicCapture();
+      return;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (rodboev) [reply static/boot.js]: Fixed in 5e5e2d58. Ctrl+Shift+D now stays enabled when focus is in the main composer textarea, while the guard still avoids stealing the shortcut from unrelated editable fields like settings inputs and dialogs. The focused regression now pins that composer-specific exception too.

**Comment by nesquena-hermes** [comment]
## 🔬 Gate certification — RED ⛔ (great push-to-talk impl; the keyboard chord is browser-reserved)

**Certified head:** `sha:5e5e2d58` (rebased onto current master, `git apply` clean) · **PR:** #5310 · _rodboev, push-to-talk hold gesture + Ctrl+Shift+D dictation shortcut (#3700)_
**Verdict:** The push-to-talk HOLD gesture is well-built and race-hardened (keep it). But the global keyboard shortcut binds **Cmd/Ctrl+Shift+D — a browser-reserved chord** (Chrome/Edge "Bookmark all tabs"), which isn't reliably `preventDefault`-able, so it fights the browser instead of giving reliable dictation. Both I and Codex flagged it. Swap the chord (or drop the shortcut, since the hold gesture is the primary #3700 ask), then ship.

### What I ran (rebased worktree `/tmp/wt-rebase-5310`)
| Gate | Result |
|------|--------|
| Rebase onto current master | ✅ merge-base == master tip (`05aa4731`); `git apply` clean |
| Codex (reproduce) | SHIP-WITH-FIXES — CORE: Cmd/Ctrl+Shift+D browser-reserved (independently confirmed my finding) |
| Full pytest suite | **11419 passed**, 0 failed (can't catch a browser-chrome chord collision) |
| Keybinding-reserved-chord check (mine) | ⛔ confirmed reserved (below) |
| Push-to-talk race review (mine) | ✅ well-hardened (below) |

### Findings
**⛔ CORE — Cmd/Ctrl+Shift+D is a browser-reserved chord** (`static/boot.js:1921`): Chrome/Edge reserve Ctrl+Shift+D for **"Bookmark all tabs"** (and Ctrl/Cmd+D = bookmark-this-page is in the reserved set; the Shift variant is the same chrome-level bookmark family as the reserved Ctrl+Shift+B). Chrome-level bookmark/tab/window chords fire before the page and are NOT reliably `preventDefault`-able, so on common setups the user gets the browser's bookmark UI (or an inconsistent fight) instead of dictation. The handler itself is clean (text-input guard except composer, `preventDefault`, then `_toggleMicCapture`) — but a perfect impl on a reserved chord is still wrong. The green suite can't catch this (it's browser-chrome behavior), and the PR's own shortcut test asserts `D` works, cementing the wrong chord.
- **Fix:** remove the Cmd/Ctrl+Shift+D branch (lines 1921-1930) OR replace it with a verified page-interceptable chord NOT in the reserved set {D, O, N, T, B, J, W, Delete} — e.g. a Cmd+K-class combo or another safe Cmd/Ctrl+Shift+<letter>. Update the shortcut test so `D` is no longer the asserted chord. (Since push-to-talk HOLD is the primary #3700 ask and works well, dropping the keyboard shortcut entirely is also a clean option.)

**✅ Keep — the push-to-talk hold gesture is well-built and race-hardened:**
- `pointerdown` → 300ms hold timer → `_startMicCapture(holdRequired=true)` only if still pressed + not already active; `pointerup` → release-to-stop if hold was active, else tap-toggles; `pointerleave` clears the timer.
- Race guard solid: `_micStartSeq` captured at start; after the async `getUserMedia`, aborts + cleans up the stream if `startSeq!==_micStartSeq || (holdRequired && !_micHoldActive)` — no stale/double-start or stuck recording (the 6 iteration commits paid off).
- i18n untouched (0 diff); the 9 mic keys the code reads pre-exist; no regression to the existing mic flow; suite green.

### Recommendation to the next agent
**RED — gate-fail/changes-requested on the chord ONLY. The gesture + hardening are ship-quality.** Contributor: swap Cmd/Ctrl+Shift+D for a non-reserved page-interceptable chord (or drop the keyboard shortcut and keep the hold gesture) + update the shortcut test. Fast converge — everything else is clean. concept 4/5 (push-to-talk dictation is a solid #3700 feature). Author @rodboev (T1). crit=2.

---
_Gate-certifier layer (warm-up → **gate** → release). I do not merge/tag/deploy. Rebased onto current master; the reserved-chord finding confirmed independently by me and Codex (Chrome/Edge bookmark-all-tabs). Cert valid for sha:5e5e2d58._

**Comment by rodboev** [comment]
I removed the `Ctrl/Cmd+Shift+D` binding instead of swapping in another browser chord on guesswork. Chrome's current shortcut docs reserve `Ctrl/Cmd+Shift+D` for bookmarking all open tabs, and Edge does the same, so the reliable fix here was to keep the hold-to-talk path and drop the global shortcut.

The new head keeps the shared mic-owner and race-hardening work intact, removes the reserved branch from `static/boot.js`, and flips `tests/test_issue3700_push_to_talk.py` into a regression that guards against reintroducing that chord.

**Comment by nesquena-hermes** [comment]
## 🔬 Gate certification — GREEN ✅ · CONVERGED (bounce closed)

**Certified head:** `sha:86f2ffe9` (clean rebase, branch `gate-rebase/5310-push-to-talk`) · **PR:** #5310 · _rodboev, push-to-talk hold gesture (#3700)_
**Verdict:** Bounce closed — my prior blocker (the browser-reserved `Cmd/Ctrl+Shift+D` chord) is fully removed, leaving the well-built, race-hardened push-to-talk hold gesture (the primary #3700 ask). Codex SAFE, suite green. Ready to merge from the rebase branch.

### What I ran (rebased worktree `/tmp/wt-rebase-5310b`)
| Gate | Result |
|------|--------|
| Rebase onto current master | ✅ merge-base == master tip (`05aa4731`); `git apply` clean |
| Codex (reproduce) | **SAFE TO SHIP** — gated the rebased worktree, 0 findings |
| Full pytest suite | **11419 passed**, 0 failed |
| Push-to-talk + mic tests | ✅ 76/76 (test_issue3700 + test_sprint20 + test_raw_audio_upload) |
| Reserved-chord re-check (mine) | ✅ 0 reserved-chord bindings remain |

### Findings
**✅ CONVERGED — the reserved chord is gone (my exact fix-spec):**
The re-push (commit "fix(#3700): drop the reserved browser dictation chord") REMOVED the global `Cmd/Ctrl+Shift+D` handler entirely. Confirmed: `grep` finds 0 Shift+D / reserved-chord dictation bindings; there is no global keyboard dictation chord left to collide with the browser. The feature is now the push-to-talk HOLD gesture alone — which is the primary #3700 ask, so dropping the shortcut is a clean resolution.

**✅ Push-to-talk hold gesture is well-built + race-hardened (kept):**
- `pointerdown` → 300ms hold timer → `_startMicCapture(holdRequired=true)` only if still pressed + not already active; `pointerup` → release-to-stop if hold was active, else tap-toggles; `pointerleave` clears the timer.
- Race guard solid: `_micStartSeq` captured at start; after async `getUserMedia`, aborts + cleans up the stream if `startSeq!==_micStartSeq || (holdRequired && !_micHoldActive)` — no stale/double-start or stuck recording.

**✅ /voice command hardening (harmless):** `cmdVoice` now checks stricter mic-visibility (`composer-control-hidden`, `aria-hidden`, computed display) before clicking — a robustness improvement, no new chord.
- i18n untouched; no regression to the mic flow; suite green.

### Recommendation to the next agent
**Ready to merge — use branch `gate-rebase/5310-push-to-talk` (`sha:86f2ffe9`), NOT the PR's stale head `49587829`.** Bounce closed: reserved chord removed, push-to-talk gesture race-hardened + test-verified (76/76), Codex SAFE + suite 11419/0. concept 4/5 (push-to-talk dictation is a solid #3700 feature). Credit @rodboev (T1, co-authored on the rebase). crit=2. Visible surface is a hold-gesture on the existing mic button (behavior test-verified; no new render/layout to screenshot).

---
_Gate-certifier layer (warm-up → **gate** → release). I do not merge/tag/deploy. Rebased onto current master; reserved-chord removal + gesture race-safety verified. Cert valid for sha:86f2ffe9._

**Comment by nesquena-hermes** [comment]
Shipped in **v0.51.812** — thanks @rodboev! 🎉

Push-to-talk hold gesture landed (closes #3700). The browser-reserved `Ctrl+Shift+D` chord from the earlier draft was dropped (collides with browser bookmark shortcuts); only the page-safe hold gesture ships, and the restart/teardown paths are race-hardened (no stuck/zombie mic). Combined-stage gate: Codex SAFE-TO-SHIP + full suite 11639 passed + 76/76 push-to-talk tests.
